core/src/main/java/io/questdb/network/AbstractIODispatcher.java

```diff
@@ -100,12 +103,13 @@ public abstract class AbstractIODispatcher<C extends IOContext<C>> extends Synch
 
     public AbstractIODispatcher(
             IODispatcherConfiguration configuration,
-            IOContextFactory<C> ioContextFactory,
-            LongGauge connectionCountGauge
+            IOContextFactory<C> ioContextFactory
     ) {
         this.LOG = LogFactory.getLog(configuration.getDispatcherLogName());
         this.configuration = configuration;
-        this.connectionCountGauge = connectionCountGauge;
+        this.connectionCountGauge = configuration.getConnectionCountGauge();
+        this.belowMaxConnectionCountCounter = configuration.getBelowMaxConnectionCountCounter();
```

Comment:
A single connection counter in addition to the existing gauge should be sufficient to fix the flakiness. The new `belowMaxConnectionCountCounter` and `aboveMaxConnectionCountCounter` counters may confuse users who want to make use of Prometheus metrics.

Reply:
i agree for the most part, except perhaps the initial problem, where dispatcher first time enters "too many connections" mode. In this case the "below max" is "1" at the very start and remains "1" when nothing can connect to the server. It could be a single counter, which counts number of "state" switched?

Reply:
Yes, my point is that we should try to make the new metrics valuable for users. In the tests we only seem to use the `belowMaxConnectionCountCounter` counter - maybe we could rename it into (total) connection counter and document that it's updated with lag (i.e. on disconnect)?

Off-topic: unfortunately I can't reproduce the flakiness locally. Do you know if it's possible to add a sleep somewhere to reproduce it reliably?

Reply:
<img width="1238" alt="image" src="https://github.com/user-attachments/assets/59fee2c7-b38a-4cfe-a5a1-602e955748fa" />

Reply:
fixed in https://github.com/questdb/questdb/pull/5276/commits/4990c10a0889ab93175bb5453854e1e871b440d0

It turns out to be a race condition, as we shutting down listener, the limit might increase, but we would never restore the listener after limit increase. Only disconnect might have a poor attempt to restore the listener. It had very dodgy condition to restore it